### PR TITLE
x86: intel64: implement runtime page tables

### LINKED ISSUE #17832: x86: update mmustructs.h and x86_mmu.c to support long mode
arch/x86/include/ia32/mmustructs.h contains all the definitions for page tables. x86_mmu.c has the code for managing page tables.

Right now, this only implements page tables for PAE on 32-bit x86. Need to add support for 64-bit page tables as well. I think the following needs to be done:

1. X86_MMU_GET_* macros should take a pdpt pointer for PAE, and a pml4 pointer for long mode
2. When long mode enabled, need a definition for PDPTEs which specify a 1GB area, instead of a page directory. Need a container union similar to union x86_mmu_pde.
3. When long mode enabled, need a X86_MMU_GET_PDPT_ADDR() macro
4. Need some ifdefs for the different fields in all the page table structure when long mode is enabled
5. Need MMU_*_MASK macros for both types of PDPT, and PML4
6. Our new coding guidelines don't allow anonymous unions. Remove all the value fields in the table entry structures, and instead create some inline functions to convert table entries into u64_t values.
7. We have lots of APIs which take a PDPT but what they really need is the toplevel pointer to all the page tables, likely need to introduce some typedef

Also need to up-level x86_mmu.c so that it works in long mode.

### PR DESCRIPTION
**Pull request by tyrantlizard77**
What's changed:

- locore area split into text, rodata, data areas for proper MMU permissions
- necessary linker symbols added to intel64 linker script to compute MMU boot regions, with page alignment if runtime page tables are enabled
- MMU_BOOT_REGION() now uses Z_STRUCT_SECTION_ITERABLE()
- Enable MMU_BOOT_REGION() section for intel64
- Bring in CONFIG_X86_MMU and some related defines to common X86 Kconfig
- Don't set ARCH_HAS_USERSPACE or ARCH_HAS_STACK_PROTECTION for intel64 yet when X86_MMU is enabled, we're not there yet
- Uplevel x86_mmu.c to common code 
- Add logic for handling PML4 table and generating PDPTs
- move z_x86_paging_init() to common kernel_arch_func.h
- Uplevel inclusion of mmustructs.h to common x86 arch.h,
  both need it for memory domain defines
- Enable runtime page tables on boot CPU and SMP CPUs as they are brought up

Currently investigating some sporadic fatal errors that seem to be related to when CPU 1 brings up a thread, but do not have a reliable way to reproduce or ideas on the root cause yet. For now, X86_MMU can only be enabled if MP_NR_CPUS=1.

Fixes: #17832

### CODE REVIEW

**Comment by andrewboie** [comment]
Passing CI results here are misleading -- a typical local sanitycheck run will have a dozen sporadic failures that simply get fixed when CI retries them.

File: arch/x86/core/intel64/locore.S [hunk 1]
```diff
@@ -622,14 +455,101 @@ IRQ(232); IRQ(233); IRQ(234); IRQ(235); IRQ(236); IRQ(237); IRQ(238); IRQ(239)
...
 /*
  * Page tables. Long mode requires them, but we don't implement any memory
  * protection yet, so these simply identity-map the first 4GB w/ 1GB pages.
  */
 
 .align 4096
 
-pml4:
+.globl z_x86_flat_ptables
+z_x86_flat_ptables:
```
Comment (gnuless) [inline arch/x86/core/intel64/locore.S]: You probably want to remove the G bit on the entries here.  Unless you explicitly invalidate these TLBs they will likely hang around on the BSP forever.  

> pdp:	.long 0x00000183		/* 0x183 = G, 1GB, R/W, P */

**Comment by gnuless** [comment]
As I mentioned in email, I suggest deferring the SMP issues to another day.  Better to prohibit the selection of SMP and memory protection in the same build for now.

File: arch/x86/core/intel64/locore.S (hunk 1 above)
Comment (andrewboie) [reply arch/x86/core/intel64/locore.S]: done.

**Comment by andrewboie** [comment]
> As I mentioned in email, I suggest deferring the SMP issues to another day. Better to prohibit the selection of SMP and memory protection in the same build for now.

Done.
I added a new "qemu_x86_long_up" target which has a single CPU and X86_MMU enabled, so that this mechanism is still tested.

**Comment by andrewboie** [comment]
> I added a new "qemu_x86_long_up" target which has a single CPU and X86_MMU enabled, so that this mechanism is still tested.

Dropped this -- it's extra CI load for something that is transient.
The restriction on X86_MMU depending on MP_NUM_CPUS=1 is in place.

**Review by dcpleung — APPROVED** [review]

**Pull request merged**